IntegerObject: Document and ensure its 32 bit

`IntegerObject` is used to set int/integer nets as defined in SystemVerilog, which are 32 bit (signed) values. See https://github.com/cocotb/cocotb/issues/268 for a discussion on signed-ness.

- `IntegerObject` doesn't check the size, which could lead to loss of data.
- We need to document that `IntegerObject` is 32 bit.
- Our backend functionality, starting with `simulator.set_signal_val_long`, uses long instead of int as data type and in the function names. That should be cleaned up to say int.

Looking into this issue and just opened an initial PR. I developed this using Modelsim (32-bit) on windows Msys2 platform. Initial compilation of the C++ libs (no local changes) revealed an issue, that is probably caused by using a 32-bit platform. I compiled this on OSX (64-bit platform) as well without any compilation warnings:
```
    cocotb/share/lib/vhpi/VhpiCbHdl.cpp: In member function 'virtual int VhpiSignalObjHdl::set_signal_value(long int, gpi_set_action_t)':
    cocotb/share/lib/vhpi/VhpiCbHdl.cpp:579:24: warning: comparison of integer expressions of different signedness: 'long int' and 'unsigned int' [-Wsign-compare]
      579 |             if ((value > EnumLimits::max()) || (value < EnumLimits::min())) {
          |                  ~~~~~~^~~~~~~~~~~~~~~~~~~
    cocotb/share/lib/vhpi/VhpiCbHdl.cpp:579:55: warning: comparison of integer expressions of different signedness: 'long int' and 'unsigned int' [-Wsign-compare]
      579 |             if ((value > EnumLimits::max()) || (value < EnumLimits::min())) {
          |                                                 ~~~~~~^~~~~~~~~~~~~~~~~~~
    cocotb/share/lib/vhpi/VhpiCbHdl.cpp:589:24: warning: comparison of integer expressions of different signedness: 'long int' and 'unsigned int' [-Wsign-compare]
      589 |             if ((value > IntLimits::max()) || (value < IntLimits::min())) {
          |                  ~~~~~~^~~~~~~~~~~~~~~~~~
    cocotb/share/lib/vhpi/VhpiCbHdl.cpp:589:54: warning: comparison of integer expressions of different signedness: 'long int' and 'unsigned int' [-Wsign-compare]
      589 |             if ((value > IntLimits::max()) || (value < IntLimits::min())) {
          |                                                ~~~~~~^~~~~~~~~~~~~~~~~~
```
This warning is relevant when changing ```simulator.set_signal_val_long``` to only accept 32-bit values. It looks like "long" is not the same size on 32-bit and 64-bit platforms. 
Interestingly, vhpi is expecting an unsigned int```typedef uint32_t vhpiIntT;``` where vpi is expecting a signed int ```typedef int             PLI_INT32;```
Should we cast the value checks VhpiCbHdl.cpp to int64_t to get rid of the warnings above? Or should we remove the value checks altogether because we would only be accepting signed 32-bit values through ```simulator.set_signal_val_long``` anyway?  
I don't have access to a vhpi simulator to see if a direct cast would cause any undesired side effects.

--- a/cocotb/handle.py
+++ b/cocotb/handle.py
@@ -45,6 +45,7 @@ _deprecation_warned = set()
 class _Limits(enum.IntEnum):
     NONE         = 0
     SIGNED_32BIT = 1  # -2**(Nbits - 1) <= value <= 2**(Nbits - 1)  where Nbits = 32
+    VECTOR_NBIT  = 2  # -2**(Nbits - 1) <= value <= 2**Nbits - 1    where Nbits <= 32
 
 
 class SimHandleBase:
@@ -715,14 +716,15 @@ class ModifiableObject(NonConstantObject):
         value, set_action = self._check_for_set_action(value)
 
         if isinstance(value, int):
+            if len(self) <= 32:
+                call_sim(self, self._handle.set_signal_val_int, set_action, value, int(_Limits.VECTOR_NBIT))
+                return
+
             min_value = -2 ** (len(self) - 1)  # minimum negative value for 'signed' number
             max_value = 2 ** len(self) - 1     # maximum positive value for 'unsigned' number
 
             if min_value <= value <= max_value:
-                if len(self) <= 32 and value < 0x7fffffff:  # set_signal_val_int() does not support full 32-bit 'unsigned' range
-                    call_sim(self, self._handle.set_signal_val_int, set_action, value, int(_Limits.NONE))
-                    return
-                elif value < 0:
+                if value < 0:
                     value = BinaryValue(value=value, n_bits=len(self), bigEndian=False, binaryRepresentation=BinaryRepresentation.TWOS_COMPLEMENT)
                 else:
                     value = BinaryValue(value=value, n_bits=len(self), bigEndian=False, binaryRepresentation=BinaryRepresentation.UNSIGNED)

--- a/cocotb/handle.py
+++ b/cocotb/handle.py
@@ -855,7 +855,7 @@ class IntegerObject(ModifiableObject):
                 "Unsupported type for integer value assignment: {} ({!r})"
                 .format(type(value), value))
 
-        call_sim(self, self._handle.set_signal_val_long, set_action, value)
+        call_sim(self, self._handle.set_signal_val_int, set_action, value)
 
     @ModifiableObject.value.getter
     def value(self) -> int:

--- a/cocotb/handle.py
+++ b/cocotb/handle.py
@@ -32,6 +32,7 @@
 import ctypes
 import warnings
 import enum
+from functools import lru_cache
 
 import cocotb
 from cocotb import simulator
@@ -43,9 +44,25 @@ _deprecation_warned = set()
 
 
 class _Limits(enum.IntEnum):
-    NONE         = 0
-    SIGNED_32BIT = 1  # -2**(Nbits - 1) <= value <= 2**(Nbits - 1)  where Nbits = 32
-    VECTOR_NBIT  = 2  # -2**(Nbits - 1) <= value <= 2**Nbits - 1    where Nbits <= 32
+    SIGNED_NBIT   = 1
+    UNSIGNED_NBIT = 2
+    VECTOR_NBIT   = 3
+
+
+@lru_cache(maxsize=None)
+def _value_limits(n_bits, limits):
+    """Calculate min/max for given number of bits and limits class"""
+    if limits == _Limits.SIGNED_NBIT:
+        min_val = -2**(n_bits-1)
+        max_val = 2**(n_bits-1) - 1
+    elif limits == _Limits.UNSIGNED_NBIT:
+        min_val = 0
+        max_val = 2**n_bits - 1
+    else:
+        min_val = -2**(n_bits-1)
+        max_val = 2**n_bits - 1
+
+    return min_val, max_val
 
 
 class SimHandleBase:
@@ -716,14 +733,12 @@ class ModifiableObject(NonConstantObject):
         value, set_action = self._check_for_set_action(value)
 
         if isinstance(value, int):
-            if len(self) <= 32:
-                call_sim(self, self._handle.set_signal_val_int, set_action, value, int(_Limits.VECTOR_NBIT))
-                return
+            min_val, max_val = _value_limits(len(self), _Limits.VECTOR_NBIT)
+            if min_val <= value <= max_val:
+                if len(self) <= 32:
+                    call_sim(self, self._handle.set_signal_val_int, set_action, value)
+                    return
 
-            min_value = -2 ** (len(self) - 1)  # minimum negative value for 'signed' number
-            max_value = 2 ** len(self) - 1     # maximum positive value for 'unsigned' number
-
-            if min_value <= value <= max_value:
                 if value < 0:
                     value = BinaryValue(value=value, n_bits=len(self), bigEndian=False, binaryRepresentation=BinaryRepresentation.TWOS_COMPLEMENT)
                 else:
@@ -845,7 +860,13 @@ class EnumObject(ModifiableObject):
                 "Unsupported type for enum value assignment: {} ({!r})"
                 .format(type(value), value))
 
-        call_sim(self, self._handle.set_signal_val_int, set_action, value, int(_Limits.NONE))
+        min_val, max_val = _value_limits(32, _Limits.UNSIGNED_NBIT)
+        if min_val <= value <= max_val:
+            call_sim(self, self._handle.set_signal_val_int, set_action, value)
+        else:
+            raise OverflowError(
+                "Int value ({!r}) out of range for assignment of enum signal ({!r})"
+                .format(value, self._name))
 
     @ModifiableObject.value.getter
     def value(self) -> int:
@@ -880,7 +901,13 @@ class IntegerObject(ModifiableObject):
                 "Unsupported type for integer value assignment: {} ({!r})"
                 .format(type(value), value))
 
-        call_sim(self, self._handle.set_signal_val_int, set_action, value, int(_Limits.SIGNED_32BIT))
+        min_val, max_val = _value_limits(32, _Limits.SIGNED_NBIT)
+        if min_val <= value <= max_val:
+            call_sim(self, self._handle.set_signal_val_int, set_action, value)
+        else:
+            raise OverflowError(
+                "Int value ({!r}) out of range for assignment of integer signal ({!r})"
+                .format(value, self._name))
 
     @ModifiableObject.value.getter
     def value(self) -> int:

--- a/cocotb/handle.py
+++ b/cocotb/handle.py
@@ -31,6 +31,7 @@
 
 import ctypes
 import warnings
+import enum
 
 import cocotb
 from cocotb import simulator
@@ -39,6 +40,11 @@ from cocotb.log import SimLog
 
 # Only issue a warning for each deprecated attribute access
 _deprecation_warned = set()
+
+
+class _Limits(enum.IntEnum):
+    NONE         = 0
+    SIGNED_32BIT = 1  # -2**(Nbits - 1) <= value <= 2**(Nbits - 1)  where Nbits = 32
 
 
 class SimHandleBase:
@@ -707,7 +713,7 @@ class ModifiableObject(NonConstantObject):
         value, set_action = self._check_for_set_action(value)
 
         if isinstance(value, int) and value < 0x7fffffff and len(self) <= 32:
-            call_sim(self, self._handle.set_signal_val_long, set_action, value)
+            call_sim(self, self._handle.set_signal_val_int, set_action, value, int(_Limits.NONE))
             return
         if isinstance(value, ctypes.Structure):
             warnings.warn(
@@ -823,7 +829,7 @@ class EnumObject(ModifiableObject):
                 "Unsupported type for enum value assignment: {} ({!r})"
                 .format(type(value), value))
 
-        call_sim(self, self._handle.set_signal_val_long, set_action, value)
+        call_sim(self, self._handle.set_signal_val_int, set_action, value, int(_Limits.NONE))
 
     @ModifiableObject.value.getter
     def value(self) -> int:
@@ -858,7 +864,7 @@ class IntegerObject(ModifiableObject):
                 "Unsupported type for integer value assignment: {} ({!r})"
                 .format(type(value), value))
 
-        call_sim(self, self._handle.set_signal_val_int, set_action, value)
+        call_sim(self, self._handle.set_signal_val_int, set_action, value, int(_Limits.SIGNED_32BIT))
 
     @ModifiableObject.value.getter
     def value(self) -> int:

--- a/cocotb/handle.py
+++ b/cocotb/handle.py
@@ -35,7 +35,7 @@ import enum
 
 import cocotb
 from cocotb import simulator
-from cocotb.binary import BinaryValue
+from cocotb.binary import BinaryValue, BinaryRepresentation
 from cocotb.log import SimLog
 
 # Only issue a warning for each deprecated attribute access
@@ -702,8 +702,10 @@ class ModifiableObject(NonConstantObject):
                 The value to drive onto the simulator object.
 
         Raises:
-            TypeError: If target is not wide enough or has an unsupported type
-                 for value assignment.
+            TypeError: If target has an unsupported type for value assignment.
+
+            OverflowError: If int value is out of the range that can be represented
+                by the target. -2**(Nbits - 1) <= value <= 2**Nbits - 1
 
         .. deprecated:: 1.5
             :class:`ctypes.Structure` objects are no longer accepted as values for assignment.
@@ -712,17 +714,29 @@ class ModifiableObject(NonConstantObject):
         """
         value, set_action = self._check_for_set_action(value)
 
-        if isinstance(value, int) and value < 0x7fffffff and len(self) <= 32:
-            call_sim(self, self._handle.set_signal_val_int, set_action, value, int(_Limits.NONE))
-            return
+        if isinstance(value, int):
+            min_value = -2 ** (len(self) - 1)  # minimum negative value for 'signed' number
+            max_value = 2 ** len(self) - 1     # maximum positive value for 'unsigned' number
+
+            if min_value <= value <= max_value:
+                if len(self) <= 32 and value < 0x7fffffff:  # set_signal_val_int() does not support full 32-bit 'unsigned' range
+                    call_sim(self, self._handle.set_signal_val_int, set_action, value, int(_Limits.NONE))
+                    return
+                elif value < 0:
+                    value = BinaryValue(value=value, n_bits=len(self), bigEndian=False, binaryRepresentation=BinaryRepresentation.TWOS_COMPLEMENT)
+                else:
+                    value = BinaryValue(value=value, n_bits=len(self), bigEndian=False, binaryRepresentation=BinaryRepresentation.UNSIGNED)
+            else:
+                raise OverflowError(
+                    "Int value ({!r}) out of range for assignment of {!r}-bit signal ({!r})"
+                    .format(value, len(self), self._name))
+
         if isinstance(value, ctypes.Structure):
             warnings.warn(
                 "`ctypes.Structure` values are no longer accepted for value assignment. "
                 "Use `BinaryValue(value=bytes(struct_obj), n_bits=len(signal))` instead",
                 DeprecationWarning, stacklevel=3)
             value = BinaryValue(value=cocotb.utils.pack(value), n_bits=len(self))
-        elif isinstance(value, int):
-            value = BinaryValue(value=value, n_bits=len(self), bigEndian=False)
         elif isinstance(value, dict):
             # We're given a dictionary with a list of values and a bit size...
             num = 0

--- a/cocotb/handle.py
+++ b/cocotb/handle.py
@@ -845,6 +845,9 @@ class IntegerObject(ModifiableObject):
         Raises:
             TypeError: If target has an unsupported type for
                  integer value assignment.
+
+            OverflowError: If value is out of range for assignment
+                 of 32-bit IntegerObject.
         """
         value, set_action = self._check_for_set_action(value)
 

--- a/cocotb/share/include/gpi.h
+++ b/cocotb/share/include/gpi.h
@@ -227,7 +227,7 @@ GPI_EXPORT int gpi_is_indexable(gpi_sim_hdl gpi_hdl);
 
 // Functions for setting the properties of a handle
 GPI_EXPORT void gpi_set_signal_value_real(gpi_sim_hdl gpi_hdl, double value, gpi_set_action_t action);
-GPI_EXPORT void gpi_set_signal_value_long(gpi_sim_hdl gpi_hdl, long value, gpi_set_action_t action);
+GPI_EXPORT void gpi_set_signal_value_int(gpi_sim_hdl gpi_hdl, int32_t value, gpi_set_action_t action);
 GPI_EXPORT void gpi_set_signal_value_binstr(gpi_sim_hdl gpi_hdl, const char *str, gpi_set_action_t action); // String of binary char(s) [1, 0, x, z]
 GPI_EXPORT void gpi_set_signal_value_str(gpi_sim_hdl gpi_hdl, const char *str, gpi_set_action_t action);    // String of ASCII char(s)
 

--- a/cocotb/share/lib/fli/FliImpl.h
+++ b/cocotb/share/lib/fli/FliImpl.h
@@ -240,7 +240,7 @@ public:
     double get_signal_value_real() override;
     long get_signal_value_long() override;
 
-    int set_signal_value(long value, gpi_set_action_t action) override;
+    int set_signal_value(int32_t value, gpi_set_action_t action) override;
     int set_signal_value(double value, gpi_set_action_t action) override;
     int set_signal_value_str(std::string &value, gpi_set_action_t action) override;
     int set_signal_value_binstr(std::string &value, gpi_set_action_t action) override;
@@ -277,7 +277,7 @@ public:
     long get_signal_value_long() override;
 
     using FliValueObjHdl::set_signal_value;
-    int set_signal_value(long value, gpi_set_action_t action) override;
+    int set_signal_value(int32_t value, gpi_set_action_t action) override;
 
     int initialise(std::string &name, std::string &fq_name) override;
 
@@ -316,7 +316,7 @@ public:
     const char* get_signal_value_binstr() override;
 
     using FliValueObjHdl::set_signal_value;
-    int set_signal_value(long value, gpi_set_action_t action) override;
+    int set_signal_value(int32_t value, gpi_set_action_t action) override;
     int set_signal_value_binstr(std::string &value, gpi_set_action_t action) override;
 
     int initialise(std::string &name, std::string &fq_name) override;
@@ -347,7 +347,7 @@ public:
     long get_signal_value_long() override;
 
     using FliValueObjHdl::set_signal_value;
-    int set_signal_value(long value, gpi_set_action_t action) override;
+    int set_signal_value(int32_t value, gpi_set_action_t action) override;
 
     int initialise(std::string &name, std::string &fq_name) override;
 };

--- a/cocotb/share/lib/fli/FliObjHdl.cpp
+++ b/cocotb/share/lib/fli/FliObjHdl.cpp
@@ -140,11 +140,11 @@ long FliValueObjHdl::get_signal_value_long()
     return -1;
 }
 
-int FliValueObjHdl::set_signal_value(long value, gpi_set_action_t action)
+int FliValueObjHdl::set_signal_value(int32_t value, gpi_set_action_t action)
 {
     COCOTB_UNUSED(value);
     COCOTB_UNUSED(action);
-    LOG_ERROR("Setting signal/variable value via long not supported for %s of type %d", m_fullname.c_str(), m_type);
+    LOG_ERROR("Setting signal/variable value via int32_t not supported for %s of type %d", m_fullname.c_str(), m_type);
     return -1;
 }
 
@@ -225,7 +225,7 @@ long FliEnumObjHdl::get_signal_value_long()
     }
 }
 
-int FliEnumObjHdl::set_signal_value(const long value, const gpi_set_action_t action)
+int FliEnumObjHdl::set_signal_value(const int32_t value, const gpi_set_action_t action)
 {
     if (action != GPI_DEPOSIT) {
         LOG_ERROR("Force or release action not supported for FLI.");
@@ -238,9 +238,9 @@ int FliEnumObjHdl::set_signal_value(const long value, const gpi_set_action_t act
     }
 
     if (m_is_var) {
-        mti_SetVarValue(get_handle<mtiVariableIdT>(), value);
+        mti_SetVarValue(get_handle<mtiVariableIdT>(), static_cast<mtiLongT>(value));
     } else {
-        mti_SetSignalValue(get_handle<mtiSignalIdT>(), value);
+        mti_SetSignalValue(get_handle<mtiSignalIdT>(), static_cast<mtiLongT>(value));
     }
 
     return 0;
@@ -315,7 +315,7 @@ const char* FliLogicObjHdl::get_signal_value_binstr()
     return m_val_buff;
 }
 
-int FliLogicObjHdl::set_signal_value(const long value, const gpi_set_action_t action)
+int FliLogicObjHdl::set_signal_value(const int32_t value, const gpi_set_action_t action)
 {
     if (action != GPI_DEPOSIT) {
         LOG_ERROR("Force or release action not supported for FLI.");
@@ -331,9 +331,9 @@ int FliLogicObjHdl::set_signal_value(const long value, const gpi_set_action_t ac
             mti_SetSignalValue(get_handle<mtiSignalIdT>(), enumVal);
         }
     } else {
-        LOG_DEBUG("set_signal_value(long)::0x%016x", value);
+        LOG_DEBUG("set_signal_value(int32_t)::0x%08x", value);
         for (int i = 0, idx = m_num_elems-1; i < m_num_elems; i++, idx--) {
-            mtiInt32T enumVal = value&(1L<<i) ? m_enum_map['1'] : m_enum_map['0'];
+            mtiInt32T enumVal = value&(1<<i) ? m_enum_map['1'] : m_enum_map['0'];
 
             m_mti_buff[idx] = (char)enumVal;
         }
@@ -432,7 +432,7 @@ long FliIntObjHdl::get_signal_value_long()
     return (long)value;
 }
 
-int FliIntObjHdl::set_signal_value(const long value, const gpi_set_action_t action)
+int FliIntObjHdl::set_signal_value(const int32_t value, const gpi_set_action_t action)
 {
     if (action != GPI_DEPOSIT) {
         LOG_ERROR("Force or release action not supported for FLI.");
@@ -440,9 +440,9 @@ int FliIntObjHdl::set_signal_value(const long value, const gpi_set_action_t acti
     }
 
     if (m_is_var) {
-        mti_SetVarValue(get_handle<mtiVariableIdT>(), value);
+        mti_SetVarValue(get_handle<mtiVariableIdT>(), static_cast<mtiLongT>(value));
     } else {
-        mti_SetSignalValue(get_handle<mtiSignalIdT>(), value);
+        mti_SetSignalValue(get_handle<mtiSignalIdT>(), static_cast<mtiLongT>(value));
     }
 
     return 0;

--- a/cocotb/share/lib/gpi/GpiCommon.cpp
+++ b/cocotb/share/lib/gpi/GpiCommon.cpp
@@ -509,7 +509,7 @@ int gpi_is_indexable(gpi_sim_hdl obj_hdl)
     return 0;
 }
 
-void gpi_set_signal_value_long(gpi_sim_hdl sig_hdl, long value, gpi_set_action_t action)
+void gpi_set_signal_value_int(gpi_sim_hdl sig_hdl, int32_t value, gpi_set_action_t action)
 {
     GpiSignalObjHdl *obj_hdl = static_cast<GpiSignalObjHdl*>(sig_hdl);
 

--- a/cocotb/share/lib/gpi/gpi_priv.h
+++ b/cocotb/share/lib/gpi/gpi_priv.h
@@ -154,7 +154,7 @@ public:
 
     int m_length = 0;
 
-    virtual int set_signal_value(const long value, gpi_set_action_t action) = 0;
+    virtual int set_signal_value(const int32_t value, gpi_set_action_t action) = 0;
     virtual int set_signal_value(const double value, gpi_set_action_t action) = 0;
     virtual int set_signal_value_str(std::string &value, gpi_set_action_t action) = 0;
     virtual int set_signal_value_binstr(std::string &value, gpi_set_action_t action) = 0;

--- a/cocotb/share/lib/simulator/simulatormodule.cpp
+++ b/cocotb/share/lib/simulator/simulatormodule.cpp
@@ -174,12 +174,6 @@ struct sim_time {
 
 static struct sim_time cache_time;
 
-enum class limits_e : int {
-    NONE = 0,
-    SIGNED_32BIT = 1,
-    VECTOR_NBIT = 2
-};
-
 /**
  * @name    Callback Handling
  * @brief   Handle a callback coming from GPI
@@ -672,37 +666,9 @@ static PyObject *set_signal_val_int(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject 
 {
     long long value;
     gpi_set_action_t action;
-    limits_e limits;
 
-    if (!PyArg_ParseTuple(args, "iLi:set_signal_val_int", &action, &value, &limits)) {
+    if (!PyArg_ParseTuple(args, "iL:set_signal_val_int", &action, &value)) {
         return NULL;
-    }
-
-    if (limits == limits_e::SIGNED_32BIT) {
-        if (value < std::numeric_limits<int32_t>::min() ||
-            value > std::numeric_limits<int32_t>::max()) {
-            const char *m_name = gpi_get_signal_name_str(self->hdl);
-            PyErr_Format(PyExc_OverflowError, "Value (%lld) out of range for assignment of 32-bit signed integer signal (%s)", value, m_name);
-            return NULL;
-        }
-    }
-    else if (limits == limits_e::VECTOR_NBIT) {
-        int m_elems = gpi_get_num_elems(self->hdl);
-
-        if (m_elems > 32) {
-            const char *m_name = gpi_get_signal_name_str(self->hdl);
-            PyErr_Format(PyExc_ValueError, "%d-bit signal (%s) is wider than 32-bit maximum supported by set_signal_val_int()", m_elems, m_name);
-            return NULL;
-        }
-
-        long long m_max = (1LL << m_elems) - 1LL;   // unsigned max value: 2**N-1
-        long long m_min = -(1LL << (m_elems - 1));  // signed min value: -2**(N-1)
-
-        if (value < m_min || value > m_max) {
-            const char *m_name = gpi_get_signal_name_str(self->hdl);
-            PyErr_Format(PyExc_OverflowError, "Value (%lld) out of range for assignment of %d-bit signal (%s)", value, m_elems, m_name);
-            return NULL;
-        }
     }
 
     gpi_set_signal_value_int(self->hdl, static_cast<int32_t>(value), action);
@@ -1198,7 +1164,7 @@ static PyMethodDef gpi_sim_hdl_methods[] = {
     },
     {"set_signal_val_int",
         (PyCFunction)set_signal_val_int, METH_VARARGS, PyDoc_STR(
-            "set_signal_val_int($self, action, value, limits, /)\n"
+            "set_signal_val_int($self, action, value, /)\n"
             "--\n\n"
             "Set the value of a signal using an int"
         )

--- a/cocotb/share/lib/simulator/simulatormodule.cpp
+++ b/cocotb/share/lib/simulator/simulatormodule.cpp
@@ -663,14 +663,14 @@ static PyObject *set_signal_val_real(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject
 
 static PyObject *set_signal_val_long(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject *args)
 {
-    long value;
+    long long value;
     gpi_set_action_t action;
 
-    if (!PyArg_ParseTuple(args, "il:set_signal_val_long", &action, &value)) {
+    if (!PyArg_ParseTuple(args, "iL:set_signal_val_long", &action, &value)) {
         return NULL;
     }
 
-    gpi_set_signal_value_long(self->hdl, value, action);
+    gpi_set_signal_value_int(self->hdl, static_cast<int32_t>(value), action);
     Py_RETURN_NONE;
 }
 

--- a/cocotb/share/lib/simulator/simulatormodule.cpp
+++ b/cocotb/share/lib/simulator/simulatormodule.cpp
@@ -661,6 +661,19 @@ static PyObject *set_signal_val_real(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject
     Py_RETURN_NONE;
 }
 
+static PyObject *set_signal_val_int(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject *args)
+{
+    int value;
+    gpi_set_action_t action;
+
+    if (!PyArg_ParseTuple(args, "ii:set_signal_val_int", &action, &value)) {
+        return NULL;
+    }
+
+    gpi_set_signal_value_int(self->hdl, static_cast<int32_t>(value), action);
+    Py_RETURN_NONE;
+}
+
 static PyObject *set_signal_val_long(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject *args)
 {
     long long value;
@@ -1167,6 +1180,13 @@ static PyMethodDef gpi_sim_hdl_methods[] = {
             "--\n\n"
             "set_signal_val_long(action: int, value: int) -> None\n"
             "Set the value of a signal using an integer.\n"
+        )
+    },
+    {"set_signal_val_int",
+        (PyCFunction)set_signal_val_int, METH_VARARGS, PyDoc_STR(
+            "set_signal_val_int($self, action, value, /)\n"
+            "--\n\n"
+            "Set the value of a signal using an int"
         )
     },
     {"set_signal_val_str",

--- a/cocotb/share/lib/simulator/simulatormodule.cpp
+++ b/cocotb/share/lib/simulator/simulatormodule.cpp
@@ -174,6 +174,11 @@ struct sim_time {
 
 static struct sim_time cache_time;
 
+enum class limits_e : int {
+    NONE = 0,
+    SIGNED_32BIT = 1
+};
+
 /**
  * @name    Callback Handling
  * @brief   Handle a callback coming from GPI
@@ -666,29 +671,19 @@ static PyObject *set_signal_val_int(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject 
 {
     long long value;
     gpi_set_action_t action;
+    limits_e limits;
 
-    if (!PyArg_ParseTuple(args, "iL:set_signal_val_int", &action, &value)) {
+    if (!PyArg_ParseTuple(args, "iLi:set_signal_val_int", &action, &value, &limits)) {
         return NULL;
     }
 
-    if (value < std::numeric_limits<int32_t>::min() ||
-        value > std::numeric_limits<int32_t>::max()) {
-        const char *m_name = gpi_get_signal_name_str(self->hdl);
-        PyErr_Format(PyExc_OverflowError, "Value (%lld) out of range for assignment of 32-bit integer signal (%s)", value, m_name);
-        return NULL;
-    }
-
-    gpi_set_signal_value_int(self->hdl, static_cast<int32_t>(value), action);
-    Py_RETURN_NONE;
-}
-
-static PyObject *set_signal_val_long(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject *args)
-{
-    long long value;
-    gpi_set_action_t action;
-
-    if (!PyArg_ParseTuple(args, "iL:set_signal_val_long", &action, &value)) {
-        return NULL;
+    if (limits == limits_e::SIGNED_32BIT) {
+        if (value < std::numeric_limits<int32_t>::min() ||
+            value > std::numeric_limits<int32_t>::max()) {
+            const char *m_name = gpi_get_signal_name_str(self->hdl);
+            PyErr_Format(PyExc_OverflowError, "Value (%lld) out of range for assignment of 32-bit signed integer signal (%s)", value, m_name);
+            return NULL;
+        }
     }
 
     gpi_set_signal_value_int(self->hdl, static_cast<int32_t>(value), action);
@@ -1182,17 +1177,9 @@ static PyMethodDef gpi_sim_hdl_methods[] = {
             "Get the value of a signal as a float."
         )
     },
-    {"set_signal_val_long",
-        (PyCFunction)set_signal_val_long, METH_VARARGS, PyDoc_STR(
-            "set_signal_val_long($self, action, value, /)\n"
-            "--\n\n"
-            "set_signal_val_long(action: int, value: int) -> None\n"
-            "Set the value of a signal using an integer.\n"
-        )
-    },
     {"set_signal_val_int",
         (PyCFunction)set_signal_val_int, METH_VARARGS, PyDoc_STR(
-            "set_signal_val_int($self, action, value, /)\n"
+            "set_signal_val_int($self, action, value, limits, /)\n"
             "--\n\n"
             "Set the value of a signal using an int"
         )

--- a/cocotb/share/lib/simulator/simulatormodule.cpp
+++ b/cocotb/share/lib/simulator/simulatormodule.cpp
@@ -176,7 +176,8 @@ static struct sim_time cache_time;
 
 enum class limits_e : int {
     NONE = 0,
-    SIGNED_32BIT = 1
+    SIGNED_32BIT = 1,
+    VECTOR_NBIT = 2
 };
 
 /**
@@ -682,6 +683,24 @@ static PyObject *set_signal_val_int(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject 
             value > std::numeric_limits<int32_t>::max()) {
             const char *m_name = gpi_get_signal_name_str(self->hdl);
             PyErr_Format(PyExc_OverflowError, "Value (%lld) out of range for assignment of 32-bit signed integer signal (%s)", value, m_name);
+            return NULL;
+        }
+    }
+    else if (limits == limits_e::VECTOR_NBIT) {
+        int m_elems = gpi_get_num_elems(self->hdl);
+
+        if (m_elems > 32) {
+            const char *m_name = gpi_get_signal_name_str(self->hdl);
+            PyErr_Format(PyExc_ValueError, "%d-bit signal (%s) is wider than 32-bit maximum supported by set_signal_val_int()", m_elems, m_name);
+            return NULL;
+        }
+
+        long long m_max = (1LL << m_elems) - 1LL;   // unsigned max value: 2**N-1
+        long long m_min = -(1LL << (m_elems - 1));  // signed min value: -2**(N-1)
+
+        if (value < m_min || value > m_max) {
+            const char *m_name = gpi_get_signal_name_str(self->hdl);
+            PyErr_Format(PyExc_OverflowError, "Value (%lld) out of range for assignment of %d-bit signal (%s)", value, m_elems, m_name);
             return NULL;
         }
     }

--- a/cocotb/share/lib/simulator/simulatormodule.cpp
+++ b/cocotb/share/lib/simulator/simulatormodule.cpp
@@ -41,6 +41,7 @@ static int sim_ending = 0;
 
 #include <cocotb_utils.h>     // COCOTB_UNUSED
 #include <type_traits>
+#include <limits>
 #include <Python.h>
 #include "gpi_logging.h"
 #include "gpi.h"
@@ -663,10 +664,17 @@ static PyObject *set_signal_val_real(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject
 
 static PyObject *set_signal_val_int(gpi_hdl_Object<gpi_sim_hdl> *self, PyObject *args)
 {
-    int value;
+    long long value;
     gpi_set_action_t action;
 
-    if (!PyArg_ParseTuple(args, "ii:set_signal_val_int", &action, &value)) {
+    if (!PyArg_ParseTuple(args, "iL:set_signal_val_int", &action, &value)) {
+        return NULL;
+    }
+
+    if (value < std::numeric_limits<int32_t>::min() ||
+        value > std::numeric_limits<int32_t>::max()) {
+        const char *m_name = gpi_get_signal_name_str(self->hdl);
+        PyErr_Format(PyExc_OverflowError, "Value (%lld) out of range for assignment of 32-bit integer signal (%s)", value, m_name);
         return NULL;
     }
 

--- a/cocotb/share/lib/vhpi/VhpiCbHdl.cpp
+++ b/cocotb/share/lib/vhpi/VhpiCbHdl.cpp
@@ -486,7 +486,7 @@ vhpiEnumT VhpiSignalObjHdl::chr2vhpi(const char value)
 }
 
 // Value related functions
-int VhpiLogicSignalObjHdl::set_signal_value(long value, gpi_set_action_t action)
+int VhpiLogicSignalObjHdl::set_signal_value(int32_t value, gpi_set_action_t action)
 {
     switch (m_value.format) {
         case vhpiEnumVal:
@@ -499,7 +499,7 @@ int VhpiLogicSignalObjHdl::set_signal_value(long value, gpi_set_action_t action)
         case vhpiLogicVecVal: {
             int i;
             for (i=0; i<m_num_elems; i++)
-                m_value.value.enumvs[m_num_elems-i-1] = value&(1L<<i) ? vhpi1 : vhpi0;
+                m_value.value.enumvs[m_num_elems-i-1] = value&(1<<i) ? vhpi1 : vhpi0;
 
             m_value.numElems = m_num_elems;
             break;
@@ -565,14 +565,14 @@ int VhpiLogicSignalObjHdl::set_signal_value_binstr(std::string &value, gpi_set_a
 }
 
 // Value related functions
-int VhpiSignalObjHdl::set_signal_value(long value, gpi_set_action_t action)
+int VhpiSignalObjHdl::set_signal_value(int32_t value, gpi_set_action_t action)
 {
     switch (m_value.format) {
         case vhpiEnumVecVal:
         case vhpiLogicVecVal: {
             int i;
             for (i=0; i<m_num_elems; i++)
-                m_value.value.enumvs[m_num_elems-i-1] = value&(1L<<i) ? vhpi1 : vhpi0;
+                m_value.value.enumvs[m_num_elems-i-1] = value&(1<<i) ? vhpi1 : vhpi0;
 
             // Since we may not get the numElems correctly from the sim and have to infer it
             // we also need to set it here as well each time.
@@ -583,21 +583,11 @@ int VhpiSignalObjHdl::set_signal_value(long value, gpi_set_action_t action)
 
         case vhpiLogicVal:
         case vhpiEnumVal: {
-            using EnumLimits = std::numeric_limits<vhpiEnumT>;
-            if ((value > EnumLimits::max()) || (value < EnumLimits::min())) {
-                LOG_ERROR("VHPI: Data loss detected");
-                return -1;
-            }
             m_value.value.enumv = static_cast<vhpiEnumT>(value);
             break;
         }
 
         case vhpiIntVal: {
-            using IntLimits = std::numeric_limits<vhpiIntT>;
-            if ((value > IntLimits::max()) || (value < IntLimits::min())) {
-                LOG_ERROR("VHPI: Data loss detected");
-                return -1;
-            }
             m_value.value.intg = static_cast<vhpiIntT>(value);
             break;
         }

--- a/cocotb/share/lib/vhpi/VhpiImpl.h
+++ b/cocotb/share/lib/vhpi/VhpiImpl.h
@@ -195,7 +195,7 @@ public:
     long get_signal_value_long() override;
 
     using GpiSignalObjHdl::set_signal_value;
-    int set_signal_value(long value, gpi_set_action_t action) override;
+    int set_signal_value(int32_t value, gpi_set_action_t action) override;
     int set_signal_value(double value, gpi_set_action_t action) override;
     int set_signal_value_str(std::string &value, gpi_set_action_t action) override;
     int set_signal_value_binstr(std::string &value, gpi_set_action_t action) override;
@@ -222,7 +222,7 @@ public:
 
 
     using GpiSignalObjHdl::set_signal_value;
-    int set_signal_value(long value, gpi_set_action_t action) override;
+    int set_signal_value(int32_t value, gpi_set_action_t action) override;
     int set_signal_value_binstr(std::string &value, gpi_set_action_t action) override;
 
     int initialise(std::string &name, std::string &fq_name) override;

--- a/cocotb/share/lib/vpi/VpiCbHdl.cpp
+++ b/cocotb/share/lib/vpi/VpiCbHdl.cpp
@@ -329,11 +329,11 @@ long VpiSignalObjHdl::get_signal_value_long()
 }
 
 // Value related functions
-int VpiSignalObjHdl::set_signal_value(long value, gpi_set_action_t action)
+int VpiSignalObjHdl::set_signal_value(int32_t value, gpi_set_action_t action)
 {
     s_vpi_value value_s;
 
-    value_s.value.integer = (int)value;
+    value_s.value.integer = static_cast<PLI_INT32>(value);
     value_s.format = vpiIntVal;
 
     return set_signal_value(value_s, action);

--- a/cocotb/share/lib/vpi/VpiImpl.h
+++ b/cocotb/share/lib/vpi/VpiImpl.h
@@ -179,7 +179,7 @@ public:
     double get_signal_value_real() override;
     long get_signal_value_long() override;
 
-    int set_signal_value(const long value, gpi_set_action_t action) override;
+    int set_signal_value(const int32_t value, gpi_set_action_t action) override;
     int set_signal_value(const double value, gpi_set_action_t action) override;
     int set_signal_value_binstr(std::string &value, gpi_set_action_t action) override;
     int set_signal_value_str(std::string &value, gpi_set_action_t action) override;

--- a/documentation/source/newsfragments/913.bugfix.rst
+++ b/documentation/source/newsfragments/913.bugfix.rst
@@ -1,0 +1,1 @@
+Assigning Python ints to signals greater than 32-bit wide will now work correctly for negative values.

--- a/documentation/source/newsfragments/913.bugfix.rst
+++ b/documentation/source/newsfragments/913.bugfix.rst
@@ -1,1 +1,1 @@
-Assigning Python ints to signals greater than 32-bit wide will now work correctly for negative values.
+Assigning Python ints to signals greater than 32 bits wide will now work correctly for negative values.

--- a/documentation/source/newsfragments/913.change.rst
+++ b/documentation/source/newsfragments/913.change.rst
@@ -1,0 +1,2 @@
+Assigning out-of-range Python ints to signals would previously truncate the value silently for signal widths <= 32 bits and truncate the value with a warning for signal widths > 32 bits.
+Assigning out-of-range Python ints to signals will now raise an :exc:`OverflowError`.

--- a/documentation/source/quickstart.rst
+++ b/documentation/source/quickstart.rst
@@ -154,6 +154,42 @@ writes are not applied immediately, but delayed until the next write cycle.
 Use ``sig.setimmediatevalue(new_val)`` to set a new value immediately
 (see :meth:`~cocotb.handle.NonHierarchyObject.setimmediatevalue`).
 
+Signed and unsigned values
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Both signed and unsigned values can be assigned to signals using a Python int.
+Cocotb makes no assumptions regarding the signedness of the signal. It only
+considers the width of the signal, so it will allow values in the range from
+the minimum negative value for a signed number up to the maximum positive
+value for an unsigned number: ``-2**(Nbits - 1) <= value <= 2**Nbits - 1``
+Note: assigning out-of-range values will raise an :exc:`OverflowError`.
+
+A :class:`BinaryValue` object can be used instead of a Python int to assign a
+value to signals with more fine-grained control (e.g. signed values only).
+
+.. code-block:: verilog
+
+    module my_module (
+        input   logic       clk,
+        input   logic       rst,
+        input   logic [2:0] data_in,
+        output  logic [2:0] data_out
+        );
+
+.. code-block:: python3
+
+    # assignment of negative value
+    dut.data_in <= -4
+
+    # assignment of positive value
+    dut.data_in <= 7
+
+    # assignment of out-of-range values
+    dut.data_in <= 8   # raises OverflowError
+    dut.data_in <= -5  # raises OverflowError
+
+Forcing and freezing signals
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 In addition to regular value assignments (deposits), signals can be forced
 to a predetermined value or frozen at their current value. To achieve this,
 the various actions described in :ref:`assignment-methods` can be used.

--- a/tests/designs/sample_module/sample_module.sv
+++ b/tests/designs/sample_module/sample_module.sv
@@ -55,7 +55,10 @@ module sample_module #(
     input  string                               stream_in_string,
 `endif
     input  [7:0]                                stream_in_data,
+    input  [31:0]                               stream_in_data_dword,
+    input  [38:0]                               stream_in_data_39bit,
     input  [63:0]                               stream_in_data_wide,
+    input  [127:0]                              stream_in_data_dqword,
 
     input                                       stream_out_ready,
     output reg [7:0]                            stream_out_data_comb,

--- a/tests/designs/sample_module/sample_module.vhdl
+++ b/tests/designs/sample_module/sample_module.vhdl
@@ -42,7 +42,10 @@ entity sample_module is
         clk                             : in    std_ulogic;
 
         stream_in_data                  : in    std_ulogic_vector(7 downto 0);
+        stream_in_data_dword            : in    std_ulogic_vector(31 downto 0);
+        stream_in_data_39bit            : in    std_ulogic_vector(38 downto 0);
         stream_in_data_wide             : in    std_ulogic_vector(63 downto 0);
+        stream_in_data_dqword           : in    std_ulogic_vector(127 downto 0);
         stream_in_valid                 : in    std_ulogic;
         stream_in_func_en               : in    std_ulogic;
         stream_in_ready                 : out   std_ulogic;

--- a/tests/test_cases/test_cocotb/test_handle.py
+++ b/tests/test_cases/test_cocotb/test_handle.py
@@ -66,6 +66,147 @@ async def test_delayed_assignment_still_errors(dut):
         dut.stream_in_int <= []
 
 
+async def int_values_test(signal, values):
+    """Test integer access to a signal."""
+
+    log = logging.getLogger("cocotb.test")
+    for val in values:
+        signal <= val
+        await Timer(10, 'ns')
+
+        if val < 0:
+            got = signal.value.signed_integer
+        else:
+            got = int(signal)
+
+        if got != val:
+            log.error("Expected value %d, got value %d" %(val, got))
+        assert got == val
+
+
+def gen_int_test_values(n_bits):
+    """Generates a list of int test values for a given number of bits."""
+    unsigned_min = 0
+    unsigned_max = 2**n_bits-1
+    signed_min = -2**(n_bits-1)
+    signed_max = 2**(n_bits-1)-1
+
+    return [1, -1, 4, -4, unsigned_min, unsigned_max, signed_min, signed_max]
+
+
+def gen_int_ovfl_value(n_bits):
+    return 2**n_bits
+
+
+def gen_int_unfl_value(n_bits):
+    return -2**(n_bits-1)-1
+
+
+@cocotb.test()
+async def test_int_8bit(dut):
+    """Test int access to 8-bit vector."""
+    values = gen_int_test_values(len(dut.stream_in_data))
+    await int_values_test(dut.stream_in_data, values)
+
+
+@cocotb.test(expect_error=OverflowError)
+async def test_int_8bit_overflow(dut):
+    """Test 8-bit vector overflow."""
+    value = gen_int_ovfl_value(len(dut.stream_in_data))
+    await int_values_test(dut.stream_in_data, [value])
+
+
+@cocotb.test(expect_error=OverflowError)
+async def test_int_8bit_underflow(dut):
+    """Test 8-bit vector underflow."""
+    value = gen_int_unfl_value(len(dut.stream_in_data))
+    await int_values_test(dut.stream_in_data, [value])
+
+
+@cocotb.test()
+async def test_int_32bit(dut):
+    """Test int access to 32-bit vector."""
+    values = gen_int_test_values(len(dut.stream_in_data_dword))
+    await int_values_test(dut.stream_in_data_dword, values)
+
+
+@cocotb.test(expect_error=OverflowError)
+async def test_int_32bit_overflow(dut):
+    """Test 32-bit vector overflow."""
+    value = gen_int_ovfl_value(len(dut.stream_in_data_dword))
+    await int_values_test(dut.stream_in_data_dword, [value])
+
+
+@cocotb.test(expect_error=OverflowError)
+async def test_int_32bit_underflow(dut):
+    """Test 32-bit vector underflow."""
+    value = gen_int_unfl_value(len(dut.stream_in_data_dword))
+    await int_values_test(dut.stream_in_data_dword, [value])
+
+
+@cocotb.test()
+async def test_int_39bit(dut):
+    """Test int access to 39-bit vector."""
+    values = gen_int_test_values(len(dut.stream_in_data_39bit))
+    await int_values_test(dut.stream_in_data_39bit, values)
+
+
+@cocotb.test(expect_error=OverflowError)
+async def test_int_39bit_overflow(dut):
+    """Test 39-bit vector overflow."""
+    value = gen_int_ovfl_value(len(dut.stream_in_data_39bit))
+    await int_values_test(dut.stream_in_data_39bit, [value])
+
+
+@cocotb.test(expect_error=OverflowError)
+async def test_int_39bit_underflow(dut):
+    """Test 39-bit vector underflow."""
+    value = gen_int_unfl_value(len(dut.stream_in_data_39bit))
+    await int_values_test(dut.stream_in_data_39bit, [value])
+
+
+@cocotb.test()
+async def test_int_64bit(dut):
+    """Test int access to 64-bit vector."""
+    values = gen_int_test_values(len(dut.stream_in_data_wide))
+    await int_values_test(dut.stream_in_data_wide, values)
+
+
+@cocotb.test(expect_error=OverflowError)
+async def test_int_64bit_overflow(dut):
+    """Test 64-bit vector overflow."""
+    value = gen_int_ovfl_value(len(dut.stream_in_data_wide))
+    await int_values_test(dut.stream_in_data_wide, [value])
+
+
+@cocotb.test(expect_error=OverflowError)
+async def test_int_64bit_underflow(dut):
+    """Test 64-bit vector underflow."""
+    value = gen_int_unfl_value(len(dut.stream_in_data_wide))
+    await int_values_test(dut.stream_in_data_wide, [value])
+
+
+@cocotb.test()
+async def test_int_128bit(dut):
+    """Test int access to 128-bit vector."""
+    values = gen_int_test_values(len(dut.stream_in_data_dqword))
+    await int_values_test(dut.stream_in_data_dqword, values)
+
+
+@cocotb.test(expect_error=OverflowError)
+async def test_int_128bit_overflow(dut):
+    """Test 128-bit vector overflow."""
+    value = gen_int_ovfl_value(len(dut.stream_in_data_dqword))
+    await int_values_test(dut.stream_in_data_dqword, [value])
+
+
+@cocotb.test(expect_error=OverflowError)
+async def test_int_128bit_underflow(dut):
+    """Test 128-bit vector underflow."""
+    value = gen_int_unfl_value(len(dut.stream_in_data_dqword))
+    await int_values_test(dut.stream_in_data_dqword, [value])
+
+
 @cocotb.test(expect_error=AttributeError if cocotb.SIM_NAME in ["Icarus Verilog"] else ())
 async def test_integer(dut):
     """Test access to integers."""

--- a/tests/test_cases/test_cocotb/test_handle.py
+++ b/tests/test_cases/test_cocotb/test_handle.py
@@ -68,19 +68,29 @@ async def test_delayed_assignment_still_errors(dut):
 
 @cocotb.test(expect_error=AttributeError if cocotb.SIM_NAME in ["Icarus Verilog"] else ())
 async def test_integer(dut):
-    """
-    Test access to integers
-    """
-    log = logging.getLogger("cocotb.test")
-    await Timer(10, "ns")
-    dut.stream_in_int = 4
-    await Timer(10, "ns")
-    await Timer(10, "ns")
-    got_in = int(dut.stream_out_int)
-    got_out = int(dut.stream_in_int)
-    log.info("dut.stream_out_int = %d" % got_out)
-    log.info("dut.stream_in_int = %d" % got_in)
-    assert got_in == got_out, "stream_in_int and stream_out_int should not match"
+    """Test access to integers."""
+    for value in [0, 1, -1, 4, -4, 2**31-1, -2**31]:
+        dut.stream_in_int = value
+        await cocotb.triggers.Timer(10, 'ns')
+        assert int(dut.stream_in_int) == value
+
+
+@cocotb.test(expect_error=AttributeError if cocotb.SIM_NAME in ["Icarus Verilog"] else OverflowError)
+async def test_integer_overflow(dut):
+    """Test integer overflow."""
+    value = 2**31
+    dut.stream_in_int = value
+    await cocotb.triggers.Timer(10, 'ns')
+    assert int(dut.stream_in_int) == value
+
+
+@cocotb.test(expect_error=AttributeError if cocotb.SIM_NAME in ["Icarus Verilog"] else OverflowError)
+async def test_integer_underflow(dut):
+    """Test integer underflow."""
+    value = -2**32-1
+    dut.stream_in_int = value
+    await cocotb.triggers.Timer(10, 'ns')
+    assert int(dut.stream_in_int) == value
 
 
 @cocotb.test(expect_error=AttributeError if cocotb.SIM_NAME in ["Icarus Verilog"] else ())

--- a/tests/test_cases/test_cocotb/test_handle.py
+++ b/tests/test_cases/test_cocotb/test_handle.py
@@ -10,6 +10,7 @@ import cocotb
 from cocotb.triggers import Timer
 
 from common import assert_raises
+from cocotb.handle import _Limits
 
 
 @cocotb.test()
@@ -66,172 +67,198 @@ async def test_delayed_assignment_still_errors(dut):
         dut.stream_in_int <= []
 
 
-async def int_values_test(signal, values):
+async def int_values_test(signal, n_bits, limits=_Limits.VECTOR_NBIT):
     """Test integer access to a signal."""
-
     log = logging.getLogger("cocotb.test")
+    values = gen_int_test_values(n_bits, limits)
     for val in values:
         signal <= val
-        await Timer(10, 'ns')
+        await Timer(1, 'ns')
 
-        if val < 0:
-            got = signal.value.signed_integer
+        if limits == _Limits.VECTOR_NBIT:
+            if val < 0:
+                got = signal.value.signed_integer
+            else:
+                got = signal.value.integer
         else:
-            got = int(signal)
+            got = signal.value
 
-        if got != val:
-            log.error("Expected value %d, got value %d" %(val, got))
-        assert got == val
+        assert got == val, "Expected value {}, got value {}!".format(val, got)
 
 
-def gen_int_test_values(n_bits):
+def gen_int_test_values(n_bits, limits=_Limits.VECTOR_NBIT):
     """Generates a list of int test values for a given number of bits."""
     unsigned_min = 0
     unsigned_max = 2**n_bits-1
     signed_min = -2**(n_bits-1)
     signed_max = 2**(n_bits-1)-1
 
-    return [1, -1, 4, -4, unsigned_min, unsigned_max, signed_min, signed_max]
+    if limits == _Limits.VECTOR_NBIT:
+        return [1, -1, 4, -4, unsigned_min, unsigned_max, signed_min, signed_max]
+    elif limits == _Limits.SIGNED_NBIT:
+        return [1, -1, 4, -4, signed_min, signed_max]
+    else:
+        return [1, -1, 4, -4, unsigned_min, unsigned_max]
 
 
-def gen_int_ovfl_value(n_bits):
-    return 2**n_bits
+async def int_overflow_test(signal, n_bits, test_mode, limits=_Limits.VECTOR_NBIT):
+    """Test integer overflow."""
+    if test_mode == "ovfl":
+        value = gen_int_ovfl_value(n_bits, limits)
+    elif test_mode == "unfl":
+        value = gen_int_unfl_value(n_bits, limits)
+    else:
+        value = None
+
+    with assert_raises(OverflowError):
+        signal <= value
 
 
-def gen_int_unfl_value(n_bits):
-    return -2**(n_bits-1)-1
+def gen_int_ovfl_value(n_bits, limits=_Limits.VECTOR_NBIT):
+    unsigned_max = 2**n_bits-1
+    signed_max = 2**(n_bits-1)-1
+
+    if limits == _Limits.SIGNED_NBIT:
+        return signed_max + 1
+    elif limits == _Limits.UNSIGNED_NBIT:
+        return unsigned_max + 1
+    else:
+        return unsigned_max + 1
+
+
+def gen_int_unfl_value(n_bits, limits=_Limits.VECTOR_NBIT):
+    unsigned_min = 0
+    signed_min = -2**(n_bits-1)
+
+    if limits == _Limits.SIGNED_NBIT:
+        return signed_min - 1
+    elif limits == _Limits.UNSIGNED_NBIT:
+        return unsigned_min - 1
+    else:
+        return signed_min - 1
 
 
 @cocotb.test()
 async def test_int_8bit(dut):
     """Test int access to 8-bit vector."""
-    values = gen_int_test_values(len(dut.stream_in_data))
-    await int_values_test(dut.stream_in_data, values)
+    await int_values_test(dut.stream_in_data, len(dut.stream_in_data))
 
 
-@cocotb.test(expect_error=OverflowError)
+@cocotb.test()
 async def test_int_8bit_overflow(dut):
     """Test 8-bit vector overflow."""
-    value = gen_int_ovfl_value(len(dut.stream_in_data))
-    await int_values_test(dut.stream_in_data, [value])
+    await int_overflow_test(dut.stream_in_data, len(dut.stream_in_data), "ovfl")
 
 
-@cocotb.test(expect_error=OverflowError)
+@cocotb.test()
 async def test_int_8bit_underflow(dut):
     """Test 8-bit vector underflow."""
-    value = gen_int_unfl_value(len(dut.stream_in_data))
-    await int_values_test(dut.stream_in_data, [value])
+    await int_overflow_test(dut.stream_in_data, len(dut.stream_in_data), "unfl")
 
 
 @cocotb.test()
 async def test_int_32bit(dut):
     """Test int access to 32-bit vector."""
-    values = gen_int_test_values(len(dut.stream_in_data_dword))
-    await int_values_test(dut.stream_in_data_dword, values)
+    await int_values_test(dut.stream_in_data_dword, len(dut.stream_in_data_dword))
 
 
-@cocotb.test(expect_error=OverflowError)
+@cocotb.test()
 async def test_int_32bit_overflow(dut):
     """Test 32-bit vector overflow."""
-    value = gen_int_ovfl_value(len(dut.stream_in_data_dword))
-    await int_values_test(dut.stream_in_data_dword, [value])
+    await int_overflow_test(dut.stream_in_data_dword, len(dut.stream_in_data_dword), "ovfl")
 
 
-@cocotb.test(expect_error=OverflowError)
+@cocotb.test()
 async def test_int_32bit_underflow(dut):
     """Test 32-bit vector underflow."""
-    value = gen_int_unfl_value(len(dut.stream_in_data_dword))
-    await int_values_test(dut.stream_in_data_dword, [value])
+    await int_overflow_test(dut.stream_in_data_dword, len(dut.stream_in_data_dword), "unfl")
 
 
 @cocotb.test()
 async def test_int_39bit(dut):
     """Test int access to 39-bit vector."""
-    values = gen_int_test_values(len(dut.stream_in_data_39bit))
-    await int_values_test(dut.stream_in_data_39bit, values)
+    await int_values_test(dut.stream_in_data_39bit, len(dut.stream_in_data_39bit))
 
 
-@cocotb.test(expect_error=OverflowError)
+@cocotb.test()
 async def test_int_39bit_overflow(dut):
     """Test 39-bit vector overflow."""
-    value = gen_int_ovfl_value(len(dut.stream_in_data_39bit))
-    await int_values_test(dut.stream_in_data_39bit, [value])
+    await int_overflow_test(dut.stream_in_data_39bit, len(dut.stream_in_data_39bit), "ovfl")
 
 
-@cocotb.test(expect_error=OverflowError)
+@cocotb.test()
 async def test_int_39bit_underflow(dut):
     """Test 39-bit vector underflow."""
-    value = gen_int_unfl_value(len(dut.stream_in_data_39bit))
-    await int_values_test(dut.stream_in_data_39bit, [value])
+    await int_overflow_test(dut.stream_in_data_39bit, len(dut.stream_in_data_39bit), "unfl")
 
 
 @cocotb.test()
 async def test_int_64bit(dut):
     """Test int access to 64-bit vector."""
-    values = gen_int_test_values(len(dut.stream_in_data_wide))
-    await int_values_test(dut.stream_in_data_wide, values)
+    await int_values_test(dut.stream_in_data_wide, len(dut.stream_in_data_wide))
 
 
-@cocotb.test(expect_error=OverflowError)
+@cocotb.test()
 async def test_int_64bit_overflow(dut):
     """Test 64-bit vector overflow."""
-    value = gen_int_ovfl_value(len(dut.stream_in_data_wide))
-    await int_values_test(dut.stream_in_data_wide, [value])
+    await int_overflow_test(dut.stream_in_data_wide, len(dut.stream_in_data_wide), "ovfl")
 
 
-@cocotb.test(expect_error=OverflowError)
+@cocotb.test()
 async def test_int_64bit_underflow(dut):
     """Test 64-bit vector underflow."""
-    value = gen_int_unfl_value(len(dut.stream_in_data_wide))
-    await int_values_test(dut.stream_in_data_wide, [value])
+    await int_overflow_test(dut.stream_in_data_wide, len(dut.stream_in_data_wide), "unfl")
 
 
 @cocotb.test()
 async def test_int_128bit(dut):
     """Test int access to 128-bit vector."""
-    values = gen_int_test_values(len(dut.stream_in_data_dqword))
-    await int_values_test(dut.stream_in_data_dqword, values)
+    await int_values_test(dut.stream_in_data_dqword, len(dut.stream_in_data_dqword))
 
 
-@cocotb.test(expect_error=OverflowError)
+@cocotb.test()
 async def test_int_128bit_overflow(dut):
     """Test 128-bit vector overflow."""
-    value = gen_int_ovfl_value(len(dut.stream_in_data_dqword))
-    await int_values_test(dut.stream_in_data_dqword, [value])
+    await int_overflow_test(dut.stream_in_data_dqword, len(dut.stream_in_data_dqword), "ovfl")
 
 
-@cocotb.test(expect_error=OverflowError)
+@cocotb.test()
 async def test_int_128bit_underflow(dut):
     """Test 128-bit vector underflow."""
-    value = gen_int_unfl_value(len(dut.stream_in_data_dqword))
-    await int_values_test(dut.stream_in_data_dqword, [value])
+    await int_overflow_test(dut.stream_in_data_dqword, len(dut.stream_in_data_dqword), "unfl")
 
 
 @cocotb.test(expect_error=AttributeError if cocotb.SIM_NAME in ["Icarus Verilog"] else ())
 async def test_integer(dut):
     """Test access to integers."""
-    for value in [0, 1, -1, 4, -4, 2**31-1, -2**31]:
-        dut.stream_in_int = value
-        await cocotb.triggers.Timer(10, 'ns')
-        assert int(dut.stream_in_int) == value
+    if cocotb.LANGUAGE in ["verilog"] and cocotb.SIM_NAME.lower().startswith("riviera"):
+        limits = _Limits.VECTOR_NBIT  # stream_in_int is ModifiableObject in riviera, not IntegerObject
+    else:
+        limits = _Limits.SIGNED_NBIT
+
+    await int_values_test(dut.stream_in_int, 32, limits)
 
 
-@cocotb.test(expect_error=AttributeError if cocotb.SIM_NAME in ["Icarus Verilog"] else OverflowError)
+@cocotb.test(expect_error=AttributeError if cocotb.SIM_NAME in ["Icarus Verilog"] else ())
 async def test_integer_overflow(dut):
     """Test integer overflow."""
-    value = 2**31
-    dut.stream_in_int = value
-    await cocotb.triggers.Timer(10, 'ns')
-    assert int(dut.stream_in_int) == value
+    if cocotb.LANGUAGE in ["verilog"] and cocotb.SIM_NAME.lower().startswith("riviera"):
+        limits = _Limits.VECTOR_NBIT  # stream_in_int is ModifiableObject in riviera, not IntegerObject
+    else:
+        limits = _Limits.SIGNED_NBIT
+
+    await int_overflow_test(dut.stream_in_int, 32, "ovfl", limits)
 
 
-@cocotb.test(expect_error=AttributeError if cocotb.SIM_NAME in ["Icarus Verilog"] else OverflowError)
+@cocotb.test(expect_error=AttributeError if cocotb.SIM_NAME in ["Icarus Verilog"] else ())
 async def test_integer_underflow(dut):
     """Test integer underflow."""
-    value = -2**32-1
-    dut.stream_in_int = value
-    await cocotb.triggers.Timer(10, 'ns')
-    assert int(dut.stream_in_int) == value
+    if cocotb.LANGUAGE in ["verilog"] and cocotb.SIM_NAME.lower().startswith("riviera"):
+        limits = _Limits.VECTOR_NBIT  # stream_in_int is ModifiableObject in riviera, not IntegerObject
+    else:
+        limits = _Limits.SIGNED_NBIT
+
+    await int_overflow_test(dut.stream_in_int, 32, "unfl", limits)
 
 
 @cocotb.test(expect_error=AttributeError if cocotb.SIM_NAME in ["Icarus Verilog"] else ())

--- a/tests/test_cases/test_multi_dimension_array/test_cocotb_array.py
+++ b/tests/test_cases/test_multi_dimension_array/test_cocotb_array.py
@@ -117,7 +117,7 @@ async def test_in_vect_packed_packed_packed(dut):
 async def test_in_vect_packed_packed_unpacked(dut):
     await Timer(1, "ns")
     print("Setting: dut.in_vect_packed_packed_unpacked type %s" % type(dut.in_vect_packed_packed_unpacked))
-    dut.in_vect_packed_packed_unpacked = [95869805, 95869805, 95869805]
+    dut.in_vect_packed_packed_unpacked = [365, 365, 365]
     await Timer(1, "ns")
     print("Getting: dut.out_vect_packed_packed_unpacked type %s" % type(dut.out_vect_packed_packed_unpacked))
     if dut.out_vect_packed_packed_unpacked != [365, 365, 365]:

